### PR TITLE
[onert] Add shape inference for onehot and its testcases.

### PR DESCRIPTION
**Pull request by Hyunjun85**
It will provide shape inference interface about onehot operations.

ONE-DCO-Signed-off-by: H Kim <hyunjun2.kim@samsung.com>

### CODE REVIEW

**Comment by Hyunjun85** [comment]
@nnfw-bot

**Comment by Hyunjun85** [comment]
@nnfw-bot test

File: runtime/onert/core/src/util/shapeinf/Onehot.cc [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  const auto indice_idx{op.getInputs().at(ir::operation::OneHot::Input::INDICES)};
+  const auto &indice = _operands.at(indice_idx);
+
+  const auto depth = op.param().depth;
+  const auto axis = op.param().axis;
+
+  auto output_idx = op.getOutputs().at(0);
+  ir::Operand &output = _operands.at(output_idx);
+
+  if (!output.info().isDynamic())
```
Comment (ragmani) [inline runtime/onert/core/src/util/shapeinf/Onehot.cc:86]: Probably, this line should be `if (!indice.info().isDynamic())`

File: runtime/onert/core/src/util/shapeinf/Onehot.cc [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+      newShape.dim(2) = features;
+    }
+    else
+    { // axis == 0
+      // new_shape will be [depth] x [batch] x [features].
+      newShape.dim(0) = depth;
+      newShape.dim(1) = batch;
+      newShape.dim(2) = features;
+    }
+  }
```
Comment (ragmani) [inline runtime/onert/core/src/util/shapeinf/Onehot.cc:70]: How about changing it like below?

```suggestion
  axis += newShape.rank();
  for (size_t i = 0; i < newShape.rank(); ++i) {
    if (i < axis) {
      newShape.dim(i) = input_shape.dim(i);
    } else if (i == axis) {
      newShape.dim(i) = depth;
    } else {
      newShape.dim(i) = input_shape.dim(i - 1); 
    }   
  }
```

File: runtime/onert/core/src/util/shapeinf/Onehot.cc [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+void DynamicInferer::visit(const ir::operation::OneHot &op)
+{
+  auto output_ind = op.getOutputs().at(0);
+  auto output = _tensor_registry->getITensor(output_ind);
+
+  auto indices_ind = op.getInputs().at(ir::operation::OneHot::INDICES);
+  const auto &indices = _tensor_registry->getITensor(indices_ind);
+  auto indices_shape = getShape(indices.get());
+
+  if (!output->is_dynamic())
```
Comment (hyunsik-yoon) [inline runtime/onert/core/src/util/shapeinf/Onehot.cc:105]: Could you check if this should be`(!input->is_dynamic)`? (please refer to other PRs).

File: runtime/onert/core/src/util/shapeinf/Onehot.cc (hunk 2 above)
Comment (Hyunjun85) [reply runtime/onert/core/src/util/shapeinf/Onehot.cc:40]: It has been updated.

File: runtime/onert/core/src/util/shapeinf/Onehot.cc (hunk 1 above)
Comment (Hyunjun85) [reply runtime/onert/core/src/util/shapeinf/Onehot.cc]: Applied.

**Comment by Hyunjun85** [comment]
This PR has been totally updated. 
I refer the reshape codes of tf-lites.

@hyunsik-yoon  @ragmani @chunseoklee 
Could you please review again?

File: tests/nnapi/specs/Ex/one_hot_2_ex_dynamic_nnfw.mod.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+model = Model()
+
+indice_input = Input("input0", "TENSOR_INT32", "{2, 2}")
+depth = Int32Scalar("depth", 3)
+onvalue = Float32Scalar("on", 5.) # default value is 1.
+offvalue = Float32Scalar("off", 1.) # default value is 0.
+axis0 = Int32Scalar("axis", 0) # default value is -1.
+model_output0 = Output("output", "TENSOR_FLOAT32", "{3, 2, 2}")
+
+model0 = model.Operation("ONE_HOT_EX", indice_input, depth, onvalue, offvalue, axis0).To(model_output0)
```
Comment (ragmani) [inline tests/nnapi/specs/Ex/one_hot_2_ex_dynamic_nnfw.mod.py:32]: This doesn't look like dynamic test.

File: tests/nnapi/specs/Ex/one_hot_1_ex_dynamic_nnfw.mod.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+#
+# If depth or axis is from input, the output shape of one_hot cannot be decided at compilation time.
+# Therefore, its output is "dynamic tensor"
+#
+
+import dynamic_tensor
+
+model = Model()
+
+indice_shape = [1]
```
Comment (ragmani) [inline tests/nnapi/specs/Ex/one_hot_1_ex_dynamic_nnfw.mod.py:27]: This is the same size as a scalar. Could you change it to a larger size?

File: tests/nnapi/specs/Ex/one_hot_2_ex_dynamic_nnfw.mod.py (hunk 4 above)
Comment (Hyunjun85) [reply tests/nnapi/specs/Ex/one_hot_2_ex_dynamic_nnfw.mod.py]: It has removed,

File: tests/nnapi/specs/Ex/one_hot_1_ex_dynamic_nnfw.mod.py (hunk 5 above)
Comment (Hyunjun85) [reply tests/nnapi/specs/Ex/one_hot_1_ex_dynamic_nnfw.mod.py]: Ok, I will.

**Comment by Hyunjun85** [comment]
Could you please help me? 
There is no existing tc for static tensor, So My  tc is first. 

Currently, I have faced a some strange things. 
The two results of one_hot is different with each other.

ex) one_hot result
**in tensorflow**
```
In [55]: x = tf.constant([1, 2, 0, 2])
In [57]: x = tf.reshape(x, [2, 2])
In [58]: y = tf.one_hot(x, 3, 1., 0., -1)
In [59]: z = sess.run(y)

In [60]: z.shape
Out[60]: (2, 2, 3)

In [61]: z
Out[61]:
array([[[0., 1., 0.],
        [0., 0., 1.]],

       [[1., 0., 0.],
        [0., 0., 1.]]], dtype=float32)
```

**in nnapi**
```
x = [1, 2, 0, 2]  x.shape = [2, 2]
depth = 3 
on_value = 1.0 off_value = 0.0
axis = -1

result shape = {2, 2, 3}
result data = {0., 0., 1.,
                      0., 1., 0., 
                      0., 0., 0., 
                      1., 0., 1.}
```

Of course, I made wrong tc. 
but, If you have a idea, could you let me know?
Thanks.

**Comment by Hyunjun85** [comment]
@nnfw-bot help

**Comment by Hyunjun85** [comment]
@nnfw-bot test all

File: tests/nnapi/specs/Ex/one_hot_2_ex_dynamic_nnfw.mod.py (hunk 4 above)
Comment (ragmani) [reply tests/nnapi/specs/Ex/one_hot_2_ex_dynamic_nnfw.mod.py]: > See the implementation of onehot, If the axis(argument) was negative value, It assigns the rank of indices. but if indices is dynamic, the rank will be 0.

As I know as if indices is dynamic, we can get shape and rank of indices in `DynamicInferer::visit()`. Why do you think the rank will be 0?

Comment (Hyunjun85) [reply tests/nnapi/specs/Ex/one_hot_2_ex_dynamic_nnfw.mod.py]: Well. I have checked that value using GDB. At that time the rank of indices is 0.
If axis > 0, axis then asix rank of indices(Now, 0).

In my think, It's good to move the axis resolving logic from OnehotLayer.configure to cker::onehot.

Could you please review again?

Comment (ragmani) [reply tests/nnapi/specs/Ex/one_hot_2_ex_dynamic_nnfw.mod.py]: I understand the problem now. Our runtime calls `OnehotLayer::configure()` in compilation time. And the runtime calls DynamicInferer::visit(OneHot) before `OnehotLayer::run()` in execution time. But `axis` was converted to positive number in `OnehotLayer::configure()` when dynamic indices's rank is 0.
Anyway, it looks good to move converting axis to `cker::OneHot()`.

**Pull request merged**